### ISSUE TITLE
`Error class: unknown class` in KMP Multi-module project

### ISSUE THREAD

**Issue by chrisbanes** [adamko-dev/dokkatoo#57]
Follow on from https://github.com/adamko-dev/dokkatoo/issues/50#issuecomment-1515977987.

Repro: https://github.com/cashapp/redwood/pull/936

Things tried:

- Checked `dokka_parameters.json` and I can't see any reference to local project dependency JARs.
- Moving to `plugins` instead of `apply plugin`. Same result.
- Workaround from #44. Unfortunately that results in `Could not create domain object 'commonMain' (DokkaSourceSetSpec)`

**Comment by aSemy** [comment]
I've done a quick hacky test and it looks like removing https://github.com/adamko-dev/dokkatoo/blob/8f735f049db75dbb3478b2a8de12f56d701d3dba/modules/dokkatoo-plugin/src/main/kotlin/adapters/DokkatooKotlinAdapter.kt#L229 fixes the problem. I'll push a fix soon and you can try the snapshot @chrisbanes

**Comment by chrisbanes** [comment]
Ha, yeah I just tried the same thing and it works as expected. Thanks for the quick response!

**Comment by chrisbanes** [comment]
Just tried out more of the built docs and while the references are now fixed, it seems that some of the links don't work (the linked docs don't exist). A key example is in `docs/build/dokkatoo/html/redwood-treehouse-composeui/app.cash.redwood.treehouse.composeui/-treehouse-content.html`. None of the docs for the classes exist.

**Comment by aSemy** [comment]
Sorry, I don't understand what you mean. Could you share a screenshot of the docs, and a link to the source code of the missing classes?

When I look in Redwood's current docs, I can see `fun TreehouseContent(...)`

<img width="1023" alt="image" src="https://user-images.githubusercontent.com/897017/233390515-d891f332-5bb8-4499-b924-87f246e990c1.png">


And the same function in the Dokkatoo generated docs 

<img width="1019" alt="image" src="https://user-images.githubusercontent.com/897017/233390633-b7ad8679-38e4-4f60-b5bf-6f2ca242a025.png">

**Comment by chrisbanes** [comment]
Try and click on those links, most will `404` (`CodeListener` for example)

**Comment by chrisbanes** [comment]
Does the Dokka plugin output a JSON file like Dokkatoo? We have a working Dokka setup so should be pretty easy to diff if it exists

**Comment by aSemy** [comment]
Interestingly if I remove the subproject logic and add explicit dependencies then the links seems to work

```
// docs/build.gradle

//rootProject.subprojects.each {
//  if (it.name == name) return
//  evaluationDependsOn(it.path)
//
//  // Each submodule needs to have the dokkatoo plugin applied
//  if (!it.plugins.hasPlugin('dev.adamko.dokkatoo-html')) return
//
//  dependencies {
//    dokkatoo(project(it.path))
//  }
//}

dependencies {
  dokkatoo projects.redwoodTreehouseComposeui
  dokkatoo projects.redwoodTreehouseHost

  // Need to manually depend on this due to https://github.com/adamko-dev/dokkatoo/issues/14
  dokkatooPluginHtml("org.jetbrains.dokka:all-modules-page-plugin:1.8.10")
  //dokkatooPluginHtml("org.jetbrains.dokka:templating-plugin:1.8.10") // shouldn't be necessary
}
```

**Comment by chrisbanes** [comment]
Ah interesting! I'll have a play around and see if I can get it to work somehow.

**Comment by aSemy** [comment]
However, if I add all dependencies manually, then clicking `CodeListener` 404's again!

```kotlin

dependencies {
//  dokkatoo projects.redwoodCli
  dokkatoo projects.redwoodCompose
  dokkatoo projects.redwoodComposeTesting
  dokkatoo projects.redwoodFlexbox
  dokkatoo projects.redwoodGradlePlugin
  dokkatoo projects.redwoodLayoutApi
  //dokkatoo projects.redwoodLayoutApiDensity
  dokkatoo projects.redwoodLayoutCompose
  dokkatoo projects.redwoodLayoutComposeui
  dokkatoo projects.redwoodLayoutDom
  dokkatoo projects.redwoodLayoutLayoutmodifiers
  dokkatoo projects.redwoodLayoutSchema
//  dokkatoo projects.redwoodLayoutSharedTest
  dokkatoo projects.redwoodLayoutTesting
  dokkatoo projects.redwoodLayoutUiview
  dokkatoo projects.redwoodLayoutView
  dokkatoo projects.redwoodLayoutWidget
  dokkatoo projects.redwoodProtocol
  dokkatoo projects.redwoodProtocolCompose
  dokkatoo projects.redwoodProtocolWidget
  dokkatoo projects.redwoodRuntime
  dokkatoo projects.redwoodSchema
//  dokkatoo projects.redwoodToolingCodegen
//  dokkatoo projects.redwoodToolingLint
//  dokkatoo projects.redwoodToolingSchema
  dokkatoo projects.redwoodTreehouse
  dokkatoo projects.redwoodTreehouseComposeui
  dokkatoo projects.redwoodTreehouseComposeuiInsets
  dokkatoo projects.redwoodTreehouseConfiguration
  dokkatoo projects.redwoodTreehouseGuest
  dokkatoo projects.redwoodTreehouseGuestCompose
  dokkatoo projects.redwoodTreehouseHost
  dokkatoo projects.redwoodTreehouseLazylayoutApi
  dokkatoo projects.redwoodTreehouseLazylayoutCompose
  dokkatoo projects.redwoodTreehouseLazylayoutComposeui
  dokkatoo projects.redwoodTreehouseLazylayoutPaging
  dokkatoo projects.redwoodTreehouseLazylayoutSchema
  dokkatoo projects.redwoodTreehouseLazylayoutTesting
  dokkatoo projects.redwoodTreehouseLazylayoutUiview
  dokkatoo projects.redwoodTreehouseLazylayoutView
  dokkatoo projects.redwoodTreehouseLazylayoutWidget
  dokkatoo projects.redwoodWidget
  dokkatoo projects.redwoodWidgetCompose

  // Need to manually depend on this due to https://github.com/adamko-dev/dokkatoo/issues/14
  dokkatooPluginHtml("org.jetbrains.dokka:all-modules-page-plugin:1.8.10")
//  dokkatooPluginHtml("org.jetbrains.dokka:templating-plugin:1.8.10")
}
```

**Comment by chrisbanes** [comment]
> Unfortunately not [Kotlin/dokka#2873](https://github.com/Kotlin/dokka/issues/2873), but it is possible to add a breakpoint in Dokka and debug into the code - but it's not very convenient

Any tips on how to debug/breakpoint this? I've tried a few things and can't get it to break. I guess diff'ing the configuration is the only way we're going to track this down.

**Comment by aSemy** [comment]
Hmm no sorry, I thought there was a thread in the Dokka issue board explaining, but I can't fin it any more. You could try asking on the [`#dokka` Slack](https://slack-chats.kotlinlang.org/c/dokka)?

Dokkatoo and Dokka should generate _identical_ content, so while even if it's not possible to check the JSON configuration, you can do a diff on `build/dokka/htmlMultiModule` vs `docs/build/dokkatoo`, or on a module basis (e.g. `redwood-treehouse-lazylayout-composeui/build/dokka-module/html` vs `redwood-treehouse-lazylayout-composeui/build/dokka/htmlPartial`).

I've done some digging and unfortunately I haven't found an answer.

One thing that I thought was interesting is that the `package-list` file that Dokka-classic generates (`build/dokka/htmlMultiModule/package-list`) is _much_ larger than the file Dokkatoo generates (`docs/build/dokkatoo/html/package-list`).


#### Dokka's package-list

```
$dokka.format:html-v1
$dokka.linkExtension:html

module:redwood-compose
app.cash.redwood.compose
module:redwood-compose-testing
app.cash.redwood.compose.testing
module:redwood-flexbox
app.cash.redwood.flexbox
module:redwood-layout-api
app.cash.redwood.layout.api
module:redwood-layout-compose
app.cash.redwood.layout.compose
module:redwood-layout-composeui
app.cash.redwood.layout.composeui
module:redwood-layout-dom
app.cash.redwood.layout.dom
module:redwood-layout-layoutmodifiers
app.cash.redwood.layout
module:redwood-layout-schema
app.cash.redwood.layout
module:redwood-layout-testing
app.cash.redwood.layout.widget
module:redwood-layout-uiview
app.cash.redwood.layout.uiview
module:redwood-layout-view
app.cash.redwood.layout.view
module:redwood-layout-widget
app.cash.redwood.layout.widget
module:redwood-protocol
app.cash.redwood.protocol
module:redwood-protocol-compose
app.cash.redwood.protocol.compose
module:redwood-protocol-widget
app.cash.redwood.protocol.widget
module:redwood-runtime
app.cash.redwood
module:redwood-treehouse
app.cash.redwood.treehouse
module:redwood-treehouse-composeui
app.cash.redwood.treehouse.composeui
module:redwood-treehouse-composeui-insets
app.cash.redwood.treehouse.composeui
module:redwood-treehouse-configuration
app.cash.redwood.treehouse
module:redwood-treehouse-guest
app.cash.redwood.treehouse
module:redwood-treehouse-guest-compose
app.cash.redwood.treehouse
module:redwood-treehouse-host
app.cash.redwood.treehouse
module:redwood-treehouse-lazylayout-api
app.cash.redwood.treehouse.lazylayout.api
module:redwood-treehouse-lazylayout-compose
app.cash.redwood.treehouse.lazylayout.compose
module:redwood-treehouse-lazylayout-composeui
app.cash.redwood.treehouse.lazylayout.composeui
module:redwood-treehouse-lazylayout-paging
app.cash.redwood.treehouse.lazylayout.paging
module:redwood-treehouse-lazylayout-schema
app.cash.redwood.treehouse.lazylayout
module:redwood-treehouse-lazylayout-testing
app.cash.redwood.treehouse.lazylayout.widget
module:redwood-treehouse-lazylayout-uiview
app.cash.redwood.treehouse.lazylayout.uiview
module:redwood-treehouse-lazylayout-view
app.cash.redwood.treehouse.lazylayout.view
module:redwood-treehouse-lazylayout-widget
app.cash.redwood.treehouse.lazylayout.widget
module:redwood-widget
app.cash.redwood.widget
module:redwood-widget-compose
app.cash.redwood.widget.compose
```


#### Dokkatoo's package-list

```
$dokka.format:html-v1
$dokka.linkExtension:html

module:redwood-compose
app.cash.redwood.compose
module:redwood-treehouse-composeui
app.cash.redwood.treehouse.composeui
module:redwood-treehouse-composeui-insets
app.cash.redwood.treehouse.composeui
module:redwood-treehouse-configuration
app.cash.redwood.treehouse
module:redwood-treehouse-guest
app.cash.redwood.treehouse
module:redwood-treehouse-guest-compose
app.cash.redwood.treehouse
module:redwood-treehouse-host
app.cash.redwood.treehouse
```

**Comment by chrisbanes** [comment]
Interesting! This might be the same issue which is blocking upgrade to Gradle 8.x: https://github.com/cashapp/redwood/pull/823

I haven't looked too deeply, but it seems that the KMP plugin does some interesting things wrt Gradle tasks, and Dokka has explicit support for KMP projects.

**Comment by aSemy** [comment]
Hmm yes, they are probably related. My guess is the error is caused by Dokka fetching the dependencies using the Kotlin Compile task

https://github.com/Kotlin/dokka/blob/dd2a38b12658e7343a1357dc5cd3850fa090a4ba/runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt#L52-L60


But Dokkatoo won't do this, because it's not compatible with Gradle config-cache.

**Comment by aSemy** [comment]
This should be fixed by #76. Redwood managed to get Dokka working, so I'll close this issue.

If there are any more improvements or problems with how Dokkatoo handles Android projects, leave a comment or make a new issue.
